File: ios/ExpoMlkitTranslationModule.swift
import ExpoModulesCore
import MLKitTranslate
import MLKitLanguageID

public class ExpoMlkitTranslationModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoMlkitTranslation")
        
        AsyncFunction("identifyLanguage") {  (text: String, promise: Promise) in
            let languageId = LanguageIdentification.languageIdentification()
            languageId.identifyLanguage(for: text) { (languageCode, error) in
                if let error = error {
                    promise.reject(error)
                    print("Failed with error: \(error)")
                    return
                }
                // If und, return nil
                if languageCode == "und" {
                    promise.resolve(nil)
                    return
                }
                promise.resolve(languageCode)
            }
        }
        
        AsyncFunction("translate") {  (text: String, source: String, target: String, promise:Promise ) in
            guard
                let sourceLanguage = TranslateLanguage.fromTag(source),
                let targetLanguage = TranslateLanguage.fromTag(target)
            else {
                promise.reject(NSError(domain: "", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid language provided, source: \(source), target: \(target)"]))
                return
            }

            let translatorOptions = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
            let translator = Translator.translator(options: translatorOptions)
            
            // Check and download models
            let conditions = ModelDownloadConditions(
                allowsCellularAccess: true,
                allowsBackgroundDownloading: true
            )
            translator.downloadModelIfNeeded(with: conditions) { error in
                guard error == nil else { return }
                
                translator.translate(text) { translatedText, error in
                    guard error == nil, let translatedText = translatedText else { return }
                    
                    // Translation succeeded.
                    promise.resolve(translatedText)
                }
            }
        }
        
        AsyncFunction("getDownloadedModels") {
            return ModelManager.modelManager().downloadedTranslateModels.map{ $0.language.rawValue }
        }
        
        
        AsyncFunction("hasDownloadedModel") { (language: String) in
            let modelsTags = ModelManager.modelManager().downloadedTranslateModels.map{ $0.language.rawValue }
            
            guard let languageTag = TranslateLanguage.fromTag(language) else {
                return false
            }
            
            return modelsTags.contains(language)
        }
    }
}

extension TranslateLanguage {
    static func fromTag(_ code: String) -> TranslateLanguage? {
        return TranslateLanguage.allLanguages().first { $0.rawValue == code }
    }
}

File: android/src/main/java/expo/modules/mlkittranslation/ExpoMlkitTranslationModule.kt
package expo.modules.mlkittranslation

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.google.mlkit.nl.languageid.LanguageIdentification
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.nl.translate.TranslateRemoteModel
import expo.modules.kotlin.exception.toCodedException
import android.util.Log
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
class ExpoMlkitTranslationModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoMlkitTranslation")

    AsyncFunction("identifyLanguage") { text: String, promise: Promise ->
      val languageIdentifier = LanguageIdentification.getClient()
      languageIdentifier.identifyLanguage(text).addOnSuccessListener { languageCode ->
        if (languageCode == "und") {
          promise.resolve(null)
          return@addOnSuccessListener
        }
        promise.resolve(languageCode)
      }.addOnFailureListener {
        promise.reject(it.toCodedException())
      }
    }

    AsyncFunction("translate") { text: String, source: String, target: String, promise: Promise ->
      val sourceLangCode = TranslateLanguage.fromLanguageTag(source)!!
      val targetLangCode = TranslateLanguage.fromLanguageTag(target)!!
        val options = TranslatorOptions.Builder()
            .setSourceLanguage(sourceLangCode)
            .setTargetLanguage(targetLangCode)
            .build()
        val translator = Translation.getClient(options)
        translator.downloadModelIfNeeded()
          .addOnSuccessListener {
              translator.translate(text)
                .addOnSuccessListener { translatedText ->
                    promise.resolve(translatedText)
                }
                .addOnFailureListener { exception ->
                    promise.reject(exception.toCodedException())
                }
          }
          .addOnFailureListener { exception ->
              promise.reject(exception.toCodedException())
          }
    }

    AsyncFunction("getDownloadedModels") { promise: Promise ->
      val modelManager = RemoteModelManager.getInstance()

      // Get translation models stored on the device.
      modelManager.getDownloadedModels(TranslateRemoteModel::class.java)
          .addOnSuccessListener { models ->
              promise.resolve(models.map { it.language })
          }
          .addOnFailureListener {
              promise.reject(it.toCodedException())
          }
    }

    AsyncFunction("hasDownloadedModel") { language: String, promise: Promise ->
      val modelManager = RemoteModelManager.getInstance()
      // Get translation models stored on the device.
      modelManager.getDownloadedModels(TranslateRemoteModel::class.java)
          .addOnSuccessListener { models ->
              val hasModel = models.any { it.language == language }
              promise.resolve(hasModel)
          }
          .addOnFailureListener {
              promise.reject(it.toCodedException())
          }
    }
  }
}